### PR TITLE
feat(docs): add fsdp design en doc

### PR DESCRIPTION
**Pull request by Major-333**
The English version of the document corresponding to this [PR](https://github.com/intelligent-machine-learning/dlrover/pull/620)

### CODE REVIEW

File: docs/design/fsdp-design-en.md [hunk 1]
```diff
@@ -0,0 +1,329 @@
+# DLRover's FSDP Elastic Training Design Document
+
+## 1. Background
+
+As large-scale machine learning models continue to grow, distributed training has become a core technology. DLRover has shown advantages in various scenarios, especially in node failure, dynamic resource allocation, and automatic recovery. While PyTorch natively supports the FSDP (Fully Sharded Data Parallel) strategy, DLRover encounters issues with dynamic resources and fault tolerance, such as OOM (Out of Memory) and the inability to reload sharded checkpoints when the world size changes.
```
Comment (workingloong) [inline docs/design/fsdp-design-en.md:5]: Shorten the length of line to 80.

File: docs/design/fsdp-design-en.md [hunk 2]
```diff
@@ -0,0 +1,373 @@
+# DLRover's FSDP Elastic Training Design Document
+
+## 1. Background
+
+Large-scale machine learning models have expanded, necessitating
```
Comment (workingloong) [inline docs/design/fsdp-design-en.md:5]: necessitating -> necessitated ?

File: docs/design/fsdp-design-en.md [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+distributed training. DLRover excels in scenarios like node failure,
+dynamic resource allocation, and automatic recovery. While PyTorch
+supports FSDP (Fully Sharded Data Parallel), DLRover faces challenges
+with dynamic resources and fault tolerance. Issues include OOM (Out of
+Memory) and difficulties reloading sharded checkpoints when world size
```
Comment (workingloong) [inline docs/design/fsdp-design-en.md:10]: The sentence has a grammar error.

Comment (Major-333) [reply docs/design/fsdp-design-en.md]: Thanks for catching those grammatical issues in the design doc. I've addressed and updated them. Please review the latest commit.

File: docs/design/fsdp-design-en.md (hunk 2 above)
Comment (Major-333) [reply docs/design/fsdp-design-en.md]: Thanks for pointing it out. I've revisited the phrasing and made refinements. The mentioned issues should be resolved now in the latest commit.

File: docs/design/fsdp-design-en.md [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+
+### 1.1 Motivation
+
+DLRover aims to support FSDP at the system level for:
+
+1. **System-Level FSDP Elastic Training:** DLRover should recognize FSDP
+training, aiding users in auto-saving and loading checkpoints.
+2. **Enhanced PyTorch FSDP Resharding:** PyTorch's save method requires
+refinement for dynamic resharding when world size varies.
+
```
Comment (merlintang) [inline docs/design/fsdp-design-en.md:21]: PyTorch's save method requires
refinement for dynamic resharding when world size varies.

save method means?

Comment (merlintang) [reply docs/design/fsdp-design-en.md:21]: world size? to word?

File: docs/design/fsdp-design-en.md [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+2. **Efficient Load/Save with Resharding:** DLRover's elasticity
+demands sharded checkpoint saves. PyTorch's current methods aren't
+optimal for save/load with resharding, especially during world size
+changes.
+
+## 2. High-Level Design
+
+Revise DLRover's `ElasticTrainer` to save/load model shards with FSDP.
+
```
Comment (merlintang) [inline docs/design/fsdp-design-en.md:40]: do we need to add the introduction for the ElasticTrainer ?

**Review by merlintang — COMMENTED** [review]
from this document, i do not know where the dlrover improve the fsdb with details.

File: docs/design/fsdp-design-en.md (hunk 4 above)
Comment (Major-333) [reply docs/design/fsdp-design-en.md:21]: Sorry for the confusion, my wording was imprecise. I'll make corrections. What I meant to say was: in order to save the model and optimizer state, we utilize the save and load functions. However, the current torch.save and torch.load in PyTorch don't support save and load in a resharding mode, which would be preferable when the world size changes.

File: docs/design/fsdp-design-en.md (hunk 5 above)
Comment (Major-333) [reply docs/design/fsdp-design-en.md:40]: OK, I will add an introduction for the ElasticTrainer.

File: docs/design/fsdp-design-en.md [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+refinement for dynamic resharding when world size varies.
+
+### 1.2 Main Challenges
+
+FSDP implementation in DLRover presents hurdles:
+
+1. **System-Level Interface for FSDP Elastic Training:** The
+`ElasticTrainer` design should simplify user interaction yet remain
+extensible. It should support easy FSDP elastic training configuration,
+custom checkpoint strategies, and shared storage paths without task
```
Comment (samplise) [inline docs/design/fsdp-design-en.md:29]: `custom` -> `customized`
Please clarify `task conflict` further

File: docs/design/fsdp-design-en.md [hunk 7]
```diff
@@ -0,0 +1,373 @@
...
+training, aiding users in auto-saving and loading checkpoints.
+2. **Enhanced PyTorch FSDP Resharding:** PyTorch's save method requires
+refinement for dynamic resharding when world size varies.
+
+### 1.2 Main Challenges
+
+FSDP implementation in DLRover presents hurdles:
+
+1. **System-Level Interface for FSDP Elastic Training:** The
+`ElasticTrainer` design should simplify user interaction yet remain
```
Comment (samplise) [inline docs/design/fsdp-design-en.md:27]: It is a little wired to say `simplify user interaction`, maybe something like `minimize users' efforts on ...`?

**Comment by workingloong** [comment]
Close it because it is outdated.

**Pull request closed without merging**